lib/mongo-milestone-db.js

```diff
@@ -0,0 +1,227 @@
+
+    this._saveMilestoneSnapshot(collectionName, snapshot)
+      .then(() => {
+        process.nextTick(callback, null);
+      })
+      .catch(error => process.nextTick(callback, error));
+  }
+
+  getMilestoneSnapshot(collectionName, id, version, callback) {
+    if (!collectionName) return process.nextTick(callback, new InvalidCollectionNameError());
+    if (!id) return process.nextTick(callback, new InvalidIdError());
+    if (!this._isValidVersion(version)) {
+      return process.nextTick(callback, new InvalidVersionError());
+    }
+
+    this._getMilestoneSnapshot(collectionName, id, version)
+      .then(snapshot => process.nextTick(callback, null, snapshot))
+      .catch(error => process.nextTick(callback, error));
+  }
+
+  _saveMilestoneSnapshot(collectionName, snapshot) {
+    return this._collection(collectionName)
+      .then((collection) => {
+        const query = {id: snapshot.id, v: snapshot.v};
+        const options = {upsert: true};
+
+        return collection.updateOne(query, snapshot, options);
+      });
+  }
+
+  _getMilestoneSnapshot(collectionName, id, version) {
+    return this._collection(collectionName)
+      .then((collection) => {
+        const query = {id: id};
+        if (version != null) {
+          query.v = {$lte: version};
+        }
+
+        return collection
+          .find(query)
+          .project({_id: 0})
+          .sort({v: -1})
+          .limit(1)
+          .next();
+      })
+      .then((snapshot) => {
+        if (snapshot) {
+          snapshot.m = snapshot.m == null ? null : snapshot.m;
+        }
+
+        return snapshot || undefined;
+      });
+  }
+
+  _close() {
+    return this._db()
+      .then((db) => {
+        this._mongoPromise = null;
+        return db.close();
+      });
+  }
+
+  _db() {
+    if (!this._mongoPromise) {
+      return Promise.reject(new MongoClosedError());
+    }
+
+    return this._mongoPromise;
+  }
+
+  _collection(collectionName) {
+    let name;
+    let collection;
+
+    return this._db()
+      .then((db) => {
+        name = MongoMilestoneDB._milestoneCollectionName(collectionName);
+        collection = db.collection(name);
+
+        if (this._shouldCreateIndex(name)) {
+          // WARNING: Creating indexes automatically like this is quite dangerous in
+          // production if we are starting with a lot of data and no indexes
+          // already. If new indexes were added or definition of these indexes were
+          // changed, users upgrading this module could unsuspectingly lock up their
+          // databases. If indexes are created as the first ops are added to a
+          // collection this won't be a problem, but this is a dangerous mechanism.
+          // Perhaps we should only warn instead of creating the indexes, especially
+          // when there is a lot of data in the collection.
+          return collection.createIndex({id: 1, v: 1}, {background: true, unique: true});
```

Comment:
I'm wondering, will it get confusing to have both `_id` and `id` properties on these Mongo documents?

One idea is to use `d`, to match what the Mongo ops collections use to indicate the Share doc id. I'm open to other ideas, or staying with `id` if we end up deciding that's fine.

Reply:
I mostly like using `id` directly because it avoids any need for mapping the snapshots when reading/writing, but if you feel strongly we can change it. I think it quickly becomes clear which `id` is which when you poke around in the data at all (if you are so inclined). And if you're poking around, you're hopefully both:

  - familiar with Mongo, and therefore unsurprised by an `_id` field containing an `ObjectID`
  - familiar with ShareDB, and therefore unsurprised by an `id` field corresponding to the ShareDB doc ID

What do you think?

Reply:
From Share PR meeting:
* Nate prefers to use `d` as the name
* See if we can use projection to rename it to `id` when reading